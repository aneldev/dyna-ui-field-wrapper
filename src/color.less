@import (reference) "~dyna-ui-styles/styles/WHITE_BLACK";
@import (reference) "~dyna-ui-styles/styles/WHITE_RED";
@import (reference) "~dyna-ui-styles/styles/WHITE_ORANGE";
@import (reference) "~dyna-ui-styles/styles/BLACK_WHITE";
@import (reference) "~dyna-ui-styles/styles/GREY_WHITE";
@import (reference) "~dyna-ui-styles/styles/ORANGE_WHITE";
@import (reference) "~dyna-ui-styles/colors/pantone-coated";

.dyna-ui-field-wrapper {
	outline: none;

	&.dyna-ui-field-wrapper {

		// common
		.dyna-ui-label {
			label{
			}
		}
		.dyna-ui-field-wrapper-container {
			.dyna-ui-field-wrapper-control-container {
				input, select {
					color: inherit;
					background-color: transparent;
				}
			}
			.dyna-ui-field-wrapper-validation-message {
			}
			.dyna-ui-field-wrapper-footer {
			}
		}


		&--color- {
			&WHITE_BLACK {
				border-color: @color-WHITE_BLACK-foreground-1;
				background-color: @color-WHITE_BLACK-background-1;

				.dyna-ui-label {

				}
				.dyna-ui-field-wrapper-container {
					background-color: @color-WHITE_BLACK-white-1;
					.dyna-ui-field-wrapper-control-container {
					}
					.dyna-ui-field-wrapper-validation-message {
						color: @color-WHITE_BLACK-error-1;
					}
					.dyna-ui-field-wrapper-footer {
					}
				}
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-field-wrapper-container {
						background-color: @color-WHITE_BLACK-background-1;
						color: @color-WHITE_BLACK-foreground-1;
					}
				}
			}
			&GREY_WHITE {
				border-color: @color-GREY_WHITE-foreground-1;
				background-color: @color-GREY_WHITE-background-2;

				.dyna-ui-label {
					border-color: @color-GREY_WHITE-foreground-1;
					background-color: @color-GREY_WHITE-background-1;
					color: @color-GREY_WHITE-foreground-1;
				}
				.dyna-ui-field-wrapper-container {
					background-color: @color-GREY_WHITE-white-1;
					color: @color-GREY_WHITE-background-1;
					.dyna-ui-field-wrapper-control-container {
					}
					.dyna-ui-field-wrapper-validation-message {
						color: @color-GREY_WHITE-error-1;
					}
					.dyna-ui-field-wrapper-footer {
					}
				}
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-field-wrapper-container {
						background-color: @color-GREY_WHITE-background-1;
						color: @color-GREY_WHITE-foreground-1;
					}
				}
			}
			&WHITE_RED {
				border-color: @color-WHITE_RED-foreground-1;

				.dyna-ui-label {
					border-color: @color-WHITE_RED-foreground-2;
					color: @color-WHITE_RED-background-1;
					background-color: @color-WHITE_RED-foreground-1;
				}
				.dyna-ui-field-wrapper-container {
					background-color: @color-WHITE_RED-white-1;
					.dyna-ui-field-wrapper-control-container {
					}
					.dyna-ui-field-wrapper-validation-message {
						color: @color-WHITE_RED-error-1;
					}
					.dyna-ui-field-wrapper-footer {
					}
				}
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-field-wrapper-container {
						background-color: @color-WHITE_RED-foreground-1;
						color: @color-WHITE_RED-background-1;
					}
				}
			}
			&BLACK_WHITE {
				border-color: @color-BLACK_WHITE-background-1;

				.dyna-ui-label {
					border-color: @color-BLACK_WHITE-background-3;
					color: @color-BLACK_WHITE-foreground-1;
					background-color: @color-BLACK_WHITE-background-1;
				}
				.dyna-ui-field-wrapper-container {
					background-color: @color-BLACK_WHITE-white-1;
					.dyna-ui-field-wrapper-control-container {
					}
					.dyna-ui-field-wrapper-validation-message {
						color: @color-BLACK_WHITE-error-1;
					}
					.dyna-ui-field-wrapper-footer {
					}
				}
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-field-wrapper-container {
						background-color: @color-BLACK_WHITE-background-1;
						color: @color-BLACK_WHITE-foreground-1;
					}
				}
			}
			&TRANSPARENT_WHITE {
				border-color: @color-BLACK_WHITE-foreground-1;

				.dyna-ui-label {
					color: @color-BLACK_WHITE-foreground-1;
				}
				.dyna-ui-field-wrapper-container {
					color: @color-BLACK_WHITE-foreground-1;
					background-color: transparent;
					.dyna-ui-field-wrapper-control-container {
					}
					.dyna-ui-field-wrapper-validation-message {
						color: @color-BLACK_WHITE-error-1;
					}
					.dyna-ui-field-wrapper-footer {
					}
				}
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-field-wrapper-container {
						background-color: transparent;
						color: @color-BLACK_WHITE-foreground-1;
					}
				}
			}
			&ORANGE_WHITE {
				border-color: @color-ORANGE_WHITE-background-1;

				.dyna-ui-label {
					border-color: @color-ORANGE_WHITE-background-3;
					background-color: @color-ORANGE_WHITE-background-1;
					color: @color-ORANGE_WHITE-foreground-1;
				}
				.dyna-ui-field-wrapper-container {
					background-color: @color-ORANGE_WHITE-white-1;
					color: black;
					.dyna-ui-field-wrapper-control-container {
					}
					.dyna-ui-field-wrapper-required {
					}
					.dyna-ui-field-wrapper-validation-message {
						color: @color-ORANGE_WHITE-error-1;
					}
					.dyna-ui-field-wrapper-footer {
					}
				}
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-field-wrapper-container {
						background-color: @color-ORANGE_WHITE-background-1;
						color: @color-ORANGE_WHITE-foreground-1;
					}
				}
			}
			&WHITE_ORANGE {
				border-color: @color-WHITE_ORANGE-foreground-1;

				.dyna-ui-label {
					border-color: @color-WHITE_ORANGE-foreground-1;
					background-color: @color-WHITE_ORANGE-background-1;
					color: @color-WHITE_ORANGE-foreground-1;
				}
				.dyna-ui-field-wrapper-container {
					border-color: @color-WHITE_ORANGE-foreground-3;
					background-color: @color-WHITE_ORANGE-white-1;
					color: @color-WHITE_ORANGE-black-1;
					.dyna-ui-field-wrapper-control-container {
					}
					.dyna-ui-field-wrapper-required {
					}
					.dyna-ui-field-wrapper-validation-message {
						color: @color-WHITE_ORANGE-error-1;
					}
					.dyna-ui-field-wrapper-footer {
					}
				}
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-field-wrapper-container {
						background-color: @color-WHITE_ORANGE-background-1;
						color: @color-WHITE_ORANGE-black-1;
					}
				}
			}
		}
	}


	// animations
	@animation-duration: 250ms;

	transition: border-color @animation-duration ease-in-out, background-color @animation-duration ease-in-out;
	.dyna-ui-label {
		transition: color @animation-duration ease-in-out, background-color @animation-duration ease-in-out;
	}
	.dyna-ui-field-wrapper-container {
		transition: color @animation-duration ease-in-out, background-color @animation-duration ease-in-out;

		.dyna-ui-field-wrapper-validation-message {
			transition: color @animation-duration ease-in-out, background-color @animation-duration ease-in-out;
		}

		.dyna-ui-field-wrapper-footer {
		}
	}

}
