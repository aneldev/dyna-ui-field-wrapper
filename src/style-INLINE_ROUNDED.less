.dyna-ui-field-wrapper {
	&.dyna-ui-field-wrapper {

		&--style-INLINE_ROUNDED {
			@border-radius: 4px;
			outline: none;
			display: flex;

			.dyna-ui-label {
				flex: 1 0 auto;
				display: flex;
				align-items: center;
				border: 1px solid;
				border-radius: @border-radius 0 0 @border-radius;
			}

			.dyna-ui-field-wrapper-container {
				flex: 1 1 100%;
				position: relative;
				border: 1px solid;
				border-left: 0;
				border-radius: 0 @border-radius @border-radius 0;

				.dyna-ui-field-wrapper-required {
					position: absolute;
					top: 4px;
					right: 16px;
				}

				.dyna-ui-field-wrapper-isLoading {
					position: absolute;
					bottom: 4px;
					right: 16px;
				}

				.dyna-ui-field-wrapper-control-container {
					width: 100%;
					> * {
						width: 100%;
					}
				}
				.dyna-ui-field-wrapper-validation-message {
					position: absolute;
				}
				.dyna-ui-field-wrapper-footer {
				}
			}

		}

		&--style-INLINE_ROUNDED&--size- {
			&XSMALL,
			&SMALL {
				.dyna-ui-label {
					padding: 12px;
					font-size: 14px;
				}
				.dyna-ui-field-wrapper-container {
					padding: 8px 12px;

					.dyna-ui-field-wrapper-required {
						font-size: 10px;
						top: 1px;
						font-weight: bold;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
						bottom: 1px;
					}

					.dyna-ui-field-wrapper-control-container {
						position: relative;
						top: 3px;
						font-size: 14px;
					}
					.dyna-ui-field-wrapper-validation-message {
						font-size: 10px;
						bottom: 1px;
						font-weight: bold;
					}
				}
			}
			&MEDIUM {
				.dyna-ui-label {
					padding: 16px;
					font-size: 14px;
				}

				.dyna-ui-field-wrapper-container {
					padding: 16px;

					.dyna-ui-field-wrapper-required {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-control-container {
						font-size: 14px;
					}

					.dyna-ui-field-wrapper-validation-message {
						font-size: 12px;
					}
				}
			}
			&LARGE {
				.dyna-ui-label {
					padding: 24px;
					font-size: 20px;
				}
				.dyna-ui-field-wrapper-container {
					padding: 24px;

					.dyna-ui-field-wrapper-required {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-control-container {
						font-size: 20px;
					}

					.dyna-ui-field-wrapper-validation-message {
						font-size: 12px;
					}
				}
			}
		}

	}
}
