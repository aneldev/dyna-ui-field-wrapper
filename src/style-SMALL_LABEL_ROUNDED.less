.dyna-ui-field-wrapper {

	&--style-SMALL_LABEL_ROUNDED {
		@border-radius: 4px;
		outline: none;

		.dyna-ui-label {
			display: inline-block;
			border: 1px solid;
			border-radius: @border-radius @border-radius 0 0;
			border-bottom: none;
		}

		.dyna-ui-field-wrapper-container {
			position: relative;
			border: 1px solid;
			border-radius: 0 @border-radius @border-radius @border-radius;

			.dyna-ui-field-wrapper-required {
				position: absolute;
				top: 4px;
				right: 16px;
			}

			.dyna-ui-field-wrapper-isLoading {
				position: absolute;
				bottom: 4px;
				right: 16px;
			}

			.dyna-ui-field-wrapper-control-container {
				width: 100%;
				> * {
					width: 100%;
				}
			}
			.dyna-ui-field-wrapper-validation-message {
				position: absolute;
			}
			.dyna-ui-field-wrapper-footer {
			}
		}
	}

	&--style-SMALL_LABEL_ROUNDED&--size- {
		&XSMALL,
		&SMALL {
			.dyna-ui-label {
				padding: 0 6px;
				font-size: 12px;
			}
			.dyna-ui-field-wrapper-container {
				padding: 8px 12px;

				.dyna-ui-field-wrapper-required {
					font-size: 10px;
					top: 1px;
					font-weight: bold;
				}

				.dyna-ui-field-wrapper-isLoading {
					font-size: 11px;
					bottom: 1px;
				}

				.dyna-ui-field-wrapper-control-container {
					font-size: 13px;
				}
				.dyna-ui-field-wrapper-validation-message {
					font-size: 10px;
					bottom: 1px;
					font-weight: bold;
				}
			}
		}
		&MEDIUM {
			.dyna-ui-label {
				padding: 2px 8px;
				font-size: 12px;
			}

			.dyna-ui-field-wrapper-container {
				padding: 16px;

				.dyna-ui-field-wrapper-required {
					font-size: 11px;
				}

				.dyna-ui-field-wrapper-isLoading {
					font-size: 11px;
				}

				.dyna-ui-field-wrapper-control-container {
					font-size: 14px;
				}

				.dyna-ui-field-wrapper-validation-message {
					font-size: 12px;
				}
			}
		}
		&LARGE {
			.dyna-ui-label {
				padding: 4px 16px;
				font-size: 16px;
			}
			.dyna-ui-field-wrapper-container {
				padding: 24px;

				.dyna-ui-field-wrapper-required {
					font-size: 11px;
				}

				.dyna-ui-field-wrapper-isLoading {
					font-size: 11px;
				}

				.dyna-ui-field-wrapper-control-container {
					font-size: 20px;
				}

				.dyna-ui-field-wrapper-validation-message {
					font-size: 12px;
				}
			}
		}
	}
}
