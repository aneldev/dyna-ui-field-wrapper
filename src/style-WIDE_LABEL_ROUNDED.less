@import (reference) "~dyna-ui-styles/styles/ORANGE_WHITE";
@import (reference) "~dyna-ui-styles/styles/RED_WHITE";
@import (reference) "~dyna-ui-styles/styles/BLACK_WHITE";
@import (reference) "~dyna-ui-styles/styles/BLACK_ORANGE";

.dyna-ui-field-wrapper {

	&.dyna-ui-field-wrapper {

		&--style-WIDE_LABEL_ROUNDED {
			@border-radius: 4px;
			outline: none;

			.dyna-ui-label {
				border: 1px solid;
				border-radius: @border-radius @border-radius 0 0;
				border-bottom-width: 0;
			}

			.dyna-ui-field-wrapper-container {
				position: relative;
				border: 1px solid;
				border-radius: 0 0 @border-radius @border-radius;

				.dyna-ui-field-wrapper-required {
					position: absolute;
					top: 4px;
					right: 16px;
				}

				.dyna-ui-field-wrapper-isLoading {
					position: absolute;
					bottom: 4px;
					right: 16px;
				}

				.dyna-ui-field-wrapper-control-container {
					width: 100%;
					> * {
						width: 100%;
					}
				}
				.dyna-ui-field-wrapper-validation-message {
					position: absolute;
				}
				.dyna-ui-field-wrapper-footer {
				}
			}
		}

		&--style-WIDE_LABEL_ROUNDED&--size- {
			&XSMALL,
			&SMALL {
				.dyna-ui-label {
					padding: 1px 6px;
					font-size: 12px;
				}
				.dyna-ui-field-wrapper-container {
					padding: 12px;

					.dyna-ui-field-wrapper-required {
						font-size: 10px;
						top: 1px;
						font-weight: bold;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
						bottom: 1px;
					}

					.dyna-ui-field-wrapper-control-container {
						font-size: 14px;
					}
					.dyna-ui-field-wrapper-validation-message {
						font-size: 10px;
						bottom: 1px;
						font-weight: bold;
					}
				}
			}
			&MEDIUM {
				.dyna-ui-label {
					padding: 2px 8px;
					font-size: 12px;
				}

				.dyna-ui-field-wrapper-container {
					padding: 16px;

					.dyna-ui-field-wrapper-required {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-control-container {
						font-size: 14px;
					}

					.dyna-ui-field-wrapper-validation-message {
						font-size: 12px;
					}
				}
			}
			&LARGE {
				.dyna-ui-label {
					padding: 4px 16px;
					font-size: 16px;
				}
				.dyna-ui-field-wrapper-container {
					padding: 24px;

					.dyna-ui-field-wrapper-required {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-control-container {
						font-size: 20px;
					}

					.dyna-ui-field-wrapper-validation-message {
						font-size: 12px;
					}
				}
			}
		}

		// color extension

		&--color- {
			&RED_WHITE {
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-label {
						border-bottom: 1px solid @color-RED_WHITE-foreground-1;
					}
				}
			}
			&ORANGE_WHITE {
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-label {
						border-bottom: 1px solid @color-ORANGE_WHITE-foreground-1;
					}
				}
			}
			&BLACK_WHITE {
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-label {
						border-bottom: 1px solid @color-BLACK_WHITE-foreground-1;
					}
				}
			}
			&BLACK_ORANGE {
				&.dyna-ui-field-wrapper--mode-VIEW {
					.dyna-ui-label {
						border-bottom: 1px solid @color-BLACK_ORANGE-foreground-1;
					}
				}
			}
		}
	}
}
