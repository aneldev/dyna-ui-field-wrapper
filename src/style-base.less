.dyna-ui-field-wrapper {
	@border-radius: 4px;

	outline: none;

	&.dyna-ui-field-wrapper {

		// commom
		.dyna-ui-label {
			label {
				margin: 0; // bootstrap clean up
			}
		}

		.dyna-ui-field-wrapper-container {
			.dyna-ui-field-wrapper-required {
				opacity: 1;
				transition: opacity 250ms ease-in-out;
			}
			.dyna-ui-field-wrapper-isLoading {
				opacity: 1;
				transition: opacity 250ms ease-in-out;
			}
			.dyna-ui-field-wrapper-control-container {
				input, select{
					color: inherit;
					background-color: transparent;
					font-size: inherit;
				}
			}
			.dyna-ui-field-wrapper-validation-message {
				opacity: 1;
				transition: opacity 250ms ease-in-out;
			}
			.dyna-ui-field-wrapper-footer {
			}
		}

		&--mode-VIEW {
			.dyna-ui-field-wrapper-container {
				.dyna-ui-field-wrapper-required,
				.dyna-ui-field-wrapper-isLoading,
				.dyna-ui-field-wrapper-validation-message {
					opacity: 0.01;
				}
				.dyna-ui-field-wrapper-control-container {
					cursor: default;
				}
			}
		}

		&--mode-EDIT {
			cursor: pointer;
			.dyna-ui-label {
				label {
					cursor: pointer;
				}
			}
		}

	}
}
