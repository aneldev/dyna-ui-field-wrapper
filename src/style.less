.dyna-ui-field-wrapper {
	@border-radius: 4px;

	outline: none;

	&.dyna-ui-field-wrapper {

		// commom
		.dyna-ui-label {
			label {
				margin: 0; // bootstrap clean up
			}
		}

		.dyna-ui-field-wrapper-container {
			.dyna-ui-field-wrapper-required {
				opacity: 1;
				transition: opacity 250ms ease-in-out;
			}
			.dyna-ui-field-wrapper-isLoading {
				opacity: 1;
				transition: opacity 250ms ease-in-out;
			}
			.dyna-ui-field-wrapper-control-container {
				input, select{
					font-size: inherit;
				}
			}
			.dyna-ui-field-wrapper-validation-message {
				opacity: 1;
				transition: opacity 250ms ease-in-out;
			}
			.dyna-ui-field-wrapper-footer {
			}
		}


		&--style- {

			&INLINE_ROUNDED {
				display: flex;
				border: 1px solid;
				border-radius: @border-radius;
				overflow: hidden;

				.dyna-ui-label {
					flex: 1 0 auto;
					border-right: 1px solid;
				}

				.dyna-ui-field-wrapper-container {
					flex: 1 1 100%;
					position: relative;

					.dyna-ui-field-wrapper-required {
						position: absolute;
						top: 4px;
						right: 16px;
					}

					.dyna-ui-field-wrapper-isLoading {
						position: absolute;
						bottom: 4px;
						right: 16px;
					}

					.dyna-ui-field-wrapper-control-container {
						width: 100%;
						> * {
							width: 100%;
						}
					}
					.dyna-ui-field-wrapper-validation-message {
						position: absolute;
					}
					.dyna-ui-field-wrapper-footer {
					}
				}

			}

		}

		&--size- {
			&XSMALL,
			&SMALL {
				.dyna-ui-label {
					padding: 12px;
					font-size: 14px;
				}
				.dyna-ui-field-wrapper-container {
					padding: 8px 12px;

					.dyna-ui-field-wrapper-required {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-control-container {
						position: relative;
						top: 3px;
						font-size: 14px;
					}
					.dyna-ui-field-wrapper-validation-message {
						font-size: 12px;
					}
				}
			}
			&MEDIUM {
				.dyna-ui-label {
					padding: 16px;
					font-size: 14px;
				}

				.dyna-ui-field-wrapper-container {
					padding: 16px;

					.dyna-ui-field-wrapper-required {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-control-container {
						font-size: 14px;
					}

					.dyna-ui-field-wrapper-validation-message {
						font-size: 12px;
					}
				}
			}
			&LARGE {
				.dyna-ui-label {
					padding: 24px;
					font-size: 20px;
				}
				.dyna-ui-field-wrapper-container {
					padding: 24px;

					.dyna-ui-field-wrapper-required {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-isLoading {
						font-size: 11px;
					}

					.dyna-ui-field-wrapper-control-container {
						font-size: 20px;
					}

					.dyna-ui-field-wrapper-validation-message {
						font-size: 12px;
					}
				}
			}
		}

		&--mode-VIEW {
			.dyna-ui-field-wrapper-container {
				.dyna-ui-field-wrapper-required,
				.dyna-ui-field-wrapper-isLoading,
				.dyna-ui-field-wrapper-validation-message {
					opacity: 0.01;
				}
				.dyna-ui-field-wrapper-control-container {
					cursor: default;
				}
			}
		}

		&--mode-EDIT {
			cursor: pointer;
			.dyna-ui-label {
				label {
					cursor: pointer;
				}
			}
		}

	}
}
