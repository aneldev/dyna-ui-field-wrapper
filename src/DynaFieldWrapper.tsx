import * as React from "react";
import {guid} from "dyna-guid";
import {EColor} from "dyna-ui-styles"

import "./style.less";
import "./color.less";

export interface IDynaFieldWrapperProps {
  className?: string;
  mode?: EMode;
  style?: EStyle;
  color?: EColor;
  size?: ESize;
  label?: TContent;
  required?: TContent;
  isLoading?: TContent;
  children: any;
  inputElementSelector?: string;  // to manipulate a HTMLInputElement
  validationMessage?: TContent;
  footer?: TContent;
  onClick?: () => void;
  onFocus?: () => void;
}

export type TContent = string | JSX.Element;

export {EColor}

export enum EMode {
  VIEW = "VIEW",
  EDIT = "EDIT",
}

export enum EStyle {
  INLINE_ROUNDED = "INLINE_ROUNDED",
}

export enum ESize {
  XSMALL = "XSMALL",
  SMALL = "SMALL",
  MEDIUM = "MEDIUM",
  LARGE = "LARGE",
}

export class DynaFieldWrapper extends React.Component<IDynaFieldWrapperProps> {
  static defaultProps: IDynaFieldWrapperProps = {
    className: null,
    mode: EMode.EDIT,
    style: EStyle.INLINE_ROUNDED,
    color: EColor.WHITE_BLACK,
    size: ESize.MEDIUM,
    label: null,
    required: null,
    isLoading: null,
    children: null,
    inputElementSelector: null,
    validationMessage: null,
    footer: null,
    onClick: () => undefined,
    onFocus: () => undefined,
  };

  private internalId: string = guid();
  private controlContainerElement: HTMLDivElement;

  private get inputElement(): HTMLInputElement {
    return this.props.inputElementSelector &&
      this.controlContainerElement &&
      this.controlContainerElement.querySelector(this.props.inputElementSelector)
      || null;
  }

  public componentDidMount() {
    if (this.props.inputElementSelector) {
      if (!this.inputElement) {
        console.error(`DynaFieldWrapper: the inputElementSelector is defined but doesn't return any input control, inputElementSelector: [${this.props.inputElementSelector}]`);
      }
      else {
        this.inputElement && this.inputElement.setAttribute('id', this.internalId);
      }
    }
  }

  private handleClick(event: MouseEvent): void {
    this.props.onClick();
  }

  private handleLabelClick(event: MouseEvent): void {
    const controlElement: HTMLInputElement = this.controlContainerElement.querySelector(this.props.inputElementSelector);
    if (controlElement) controlElement.focus();
    this.props.onFocus();
  }

  private handleContainerClick(event: MouseEvent): void {
    const controlElement: HTMLInputElement = this.controlContainerElement.querySelector(this.props.inputElementSelector);
    if (event.target !== event.currentTarget) return;
    if (controlElement) controlElement.focus();
    this.props.onFocus();
  }

  public render(): JSX.Element {
    const {
      className: cn,
      mode, style, color, size,
      label, required, isLoading,
      children,
      validationMessage, footer,
    } = this.props;

    const className: string = [
      cn || '',
      'dyna-ui-field-wrapper',
      `dyna-ui-field-wrapper--mode-${mode}`,
      `dyna-ui-field-wrapper--style-${style}`,
      `dyna-ui-field-wrapper--color-${color}`,
      `dyna-ui-field-wrapper--size-${size}`,
    ].join(' ').trim();

    return (
      <div className={className} onClick={this.handleClick.bind(this)}>
        {label ?
          <div className="dyna-ui-label" onClick={this.handleLabelClick.bind(this)}>
            <label htmlFor={this.internalId} onClick={e=>e.stopPropagation()} >{label}</label>
          </div>
          : null}
        <div className="dyna-ui-field-wrapper-container" onClick={this.handleContainerClick.bind(this)}>
          <div className="dyna-ui-field-wrapper-required" onClick={this.handleContainerClick.bind(this)}>{required}</div>
          <div className="dyna-ui-field-wrapper-isLoading" onClick={this.handleContainerClick.bind(this)}>{isLoading}</div>
          <div className="dyna-ui-field-wrapper-control-container" ref={element => this.controlContainerElement = element} onClick={this.handleContainerClick.bind(this)}>
            {children}
          </div>
          <div className="dyna-ui-field-wrapper-validation-message" onClick={this.handleContainerClick.bind(this)}>{validationMessage}</div>
          <div className="dyna-ui-field-wrapper-footer" onClick={this.handleContainerClick.bind(this)}>{footer}</div>
        </div>
      </div>
    );
  }
}
