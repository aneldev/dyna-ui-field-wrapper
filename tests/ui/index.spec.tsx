import {configure} from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

import * as React from 'react';
import * as enzyme from 'enzyme';

import {DynaFieldWrapper} from '../../src';

describe('Home', () => {
  let wrapper;

  it('has expected content with deep render', () => {
    wrapper = enzyme.shallow(
      (
        <DynaFieldWrapper className="my-custom-class-name">
          <input className="input-control" style={{width: "100%"}} value="John" onChange={(e) => console.log('changed', e.target.value)}/>
        </DynaFieldWrapper>
      ),
      {}
    );

    expect(wrapper).toMatchSnapshot();
  });
});
