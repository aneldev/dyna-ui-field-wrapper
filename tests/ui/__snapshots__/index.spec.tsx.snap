// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home has expected content with deep render 1`] = `
<div
  className="my-custom-class-name dyna-ui-field-wrapper dyna-ui-field-wrapper-style-INLINE_ROUNDED dyna-ui-field-wrapper-color-WHITE_BLACK"
>
  <div
    className="dyna-ui-field-wrapper-container"
    onClick={[Function]}
  >
    <div
      className="dyna-ui-field-wrapper-required"
      onClick={[Function]}
    />
    <div
      className="dyna-ui-field-wrapper-isLoading"
      onClick={[Function]}
    />
    <div
      className="dyna-ui-field-wrapper-control-container"
      onClick={[Function]}
    >
      <input
        className="input-control"
        onChange={[Function]}
        style={
          Object {
            "width": "100%",
          }
        }
        value="John"
      />
    </div>
    <div
      className="dyna-ui-field-wrapper-validation-message"
      onClick={[Function]}
    />
    <div
      className="dyna-ui-field-wrapper-footer"
      onClick={[Function]}
    />
  </div>
</div>
`;
